noclip.website appears broken in Safari 18

mwyrzykowski opened the pull request:
#### 2aa166d7c00b9648ba05078375bdeb4f965304b4
<pre>
noclip.website appears broken in Safari 18
<a href="https://bugs.webkit.org/show_bug.cgi?id=280799">https://bugs.webkit.org/show_bug.cgi?id=280799</a>
<a href="https://rdar.apple.com/137643297">rdar://137643297</a>

Reviewed by Ada Chan.

Partial revert of 280236@main to so invalidateFramebuffer()
works as expected.

* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
(gl::Framebuffer::discard):
(gl::Framebuffer::invalidate):
(gl::Framebuffer::markDrawAttachmentsNeedInit): Deleted.
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):

Canonical link: <a href="https://commits.webkit.org/286127@main">https://commits.webkit.org/286127@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb2c24154ec3c1fac59adc0c9fd346eac15084b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74734 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54164 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79156 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25972 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63300 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1949 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58701 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16987 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48854 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64223 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39095 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46084 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21720 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24305 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67290 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80647 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2052 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1216 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66960 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2200 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64241 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66251 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16495 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10198 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8350 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2017 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4804 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2045 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/2965 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2052 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
```diff
@@ -1812,8 +1810,6 @@ angle::Result Framebuffer::invalidate(const Context *context,
     // Back-ends might make the contents of the FBO undefined. In WebGL 2.0, invalidate operations
     // can be no-ops, so we should probably do that to ensure consistency.
     // TODO(jmadill): WebGL behaviour, and robust resource init behaviour without WebGL.
-    if (context->isRobustResourceInitEnabled())
-        markDrawAttachmentsNeedInit(count, attachments);
```
This will break some WebXR content. An explicit clear will need to be added back into WebXROpaqueFramebuffer too.

mwyrzykowski replied:
Maybe we can mark the attachment WebXROpaqueFramebuffer uses as needing to take this path, as there is the expectation the clear occurs there, without adding the manual clear back?

I will give that a try

mwyrzykowski replied:
Is it valid to only call `markDrawAttachmentsNeedInit` from the `discardFramebuffer` path (which WebXR uses)?

I updated the patch to do that to avoid causing a performance regression in WebXR content. Maybe there is a better way to do it.

djg replied:
> Is it valid to only call markDrawAttachmentsNeedInit from the discardFramebuffer path (which WebXR uses)?

That was the intention. Internally, can we verify that the discard path is only used for WebXR and not for the WebGL2 Invalidate framebuffer?

It would be safer to put an explicit glClear into the `WebXROpaqueFramebuffer::startFrame`, or modify ANGLE to replace `MTLLoadActionDontCare` with `MTLLoadActionClear` when robust initialisation is enabled on the context. (This is my preferred solution)

mwyrzykowski replied:
Sounds good, I will do that instead

mwyrzykowski replied:
Maybe I don't understand the proposal, but just using `MTLLoadActionClear` still results in the same issue.

Let me check a gputrace to see where this is failing

mwyrzykowski committed:
noclip.website appears broken in Safari 18
https://bugs.webkit.org/show_bug.cgi?id=280799
rdar://137643297

Reviewed by Ada Chan.

Partial revert of 280236@main to so invalidateFramebuffer()
works as expected.

* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.cpp:
(gl::Framebuffer::discard):
(gl::Framebuffer::invalidate):
(gl::Framebuffer::markDrawAttachmentsNeedInit): Deleted.
* Source/ThirdParty/ANGLE/src/libANGLE/Framebuffer.h:
* Source/WebCore/Modules/webxr/WebXROpaqueFramebufferCocoa.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):

Canonical link: https://commits.webkit.org/286127@main
